### PR TITLE
i/builtin/xilinx-dma: add interface for Xilinx DMA driver

### PR DESCRIPTION
**Pull request by alexclewontin**
Needed for a customer enablement project.

Should be a fairly self-explanatory hardware access interface. The hardware in question is a Xilinx FPGA PCIe card. This interface is for use on the device that is hosting the PCIe card, and allows it access to the character device streams provided by the driver (from https://github.com/Xilinx/dma_ip_drivers)

### CODE REVIEW

**Comment by anonymouse64** [comment]
Seems reasonable, but also seems like a good candidate for the "raw-device" interface we have been designing/planning for a while, though that is still not ready yet unfortunately

**Comment by alexclewontin** [comment]
@anonymouse64 Anything I can do to help this along? We are reaching the latter stages of the enablement project this is needed for

**Review by anonymouse64 — APPROVED** [review]
lgtm, I think we need Samuele to decide the name on this, if it should be `xilinx-dma-host` or `xdma-host` or `xilinx-dma` etc. I'm not sure the host is necessary since it's kinda implied that the DMA is "on the host" from the perspective of the snap. There also doesn't seem to be much referencing "host" in the linked PDF or github page either

**Comment by alexclewontin** [comment]
> lgtm, I think we need Samuele to decide the name on this, if it should be `xilinx-dma-host` or `xdma-host` or `xilinx-dma` etc. I'm not sure the host is necessary since it's kinda implied that the DMA is "on the host" from the perspective of the snap. There also doesn't seem to be much referencing "host" in the linked PDF or github page either 

Sounds good. I am not attached to any particular name and will happily defer to you all. Just to document my reasoning for the inital draft: 

- `xilinx-dma` instead of `xdma` because there are a couple of related but distinct drivers (`qdma` and `xvsec`) that I figured might eventually need to get included. Probably not as important as they would be covered by the generic device in the future.

- `host` was my own terminology, the idea being that there could be a set up with an UltraScale+ or the like running Linux, on the other (recieving) end the DMA. I don't know how this would look from that perspective, or if that's even a realistic setup. Also maybe less important in light of the generic device interface.

I may have not been super clear about the DMA terminology. The memory being directly accessed is not owned by the host, but rather by the device on the PCIe card. The PCIe host is writing to/reading from that memory.

File: interfaces/builtin/xilinx_dma_host.go
Comment (pedronis) [inline interfaces/builtin/xilinx_dma_host.go:55]: is there a strong reason to add the "-host" suffix to this name? do we imagine running on the fpga card itself?

**Review by pedronis — COMMENTED** [review]
naming question

File: interfaces/builtin/xilinx_dma_host.go
Comment (alexclewontin) [reply interfaces/builtin/xilinx_dma_host.go:55]: Not for this project in particular. It is theoretically possible, but I have no idea what it would look like from perspective of the kernel running on the card (what driver it would use, what device nodes it would set up, etc)

Comment (alexclewontin) [reply interfaces/builtin/xilinx_dma_host.go:55]: Happy to drop the "-host" suffix if you think it makes more sense that way.

Comment (pedronis) [reply interfaces/builtin/xilinx_dma_host.go:55]: yes, my suggestion would be to drop it, also because "host" in a snap context tends to refer to the unconfined and corresponding file system outside of snap mount namespaces.

Comment (alexclewontin) [reply interfaces/builtin/xilinx_dma_host.go]: dropped

**Review by pedronis — COMMENTED** [review]
thanks, this needs a security review now

**Review by pedronis — APPROVED** [review]

**Review by alexmurray — APPROVED** [review]
So in general this looks fine to me - however I wonder whether the interface may need to be super-privileged? I am having some trouble trying to understand from the referenced documentation whether say the hardware can handle multiple concurrent users / snaps accessing it, or whether say one snap could misconfigure it to then stop other snaps from using it etc. If this is the case then I think we should make the interface super-privileged (such that a snap store declaration is required to be able to plug the interface).

However, if this is not likely to be a problem then the interface should be fine as is, so I have given my tentative approval but would like to have some comment on the above.

**Comment by alexclewontin** [comment]
@alexmurray I think the short answer is yes. 

The longer answer is, in my understanding, that the `/dev/xdma[0-9]*_{c2h,h2c,events}_[0-9]*` are in general meant to be discrete channels that can be used in parallel and access to them (likely, see caveat) wouldn't present a problem for multiple snaps. On the other hand, `/dev/xdma[0-9]*_{control,user,xvc}` are meant to be global config parameters, of the sort that could get misconfigured by multiple snaps competing for them. 

The other caveat is that these channels simply carry arbitrary data to the corresponding IP core, and how that data is handled depends on what other programmable logic the user wires it up to.

In any case, I don't think it would hurt to make it superpriviliged.

**Comment by alexmurray** [comment]
So the only 'hurt' in making this super-privileged is that snaps which wish to use this now need a store declaration - so either the respective brand store admin has to create one, or the reviewers team would have to for the global snap store. So this can hinder adoption/use of a given interface due to this increased friction. However, I suspect your team would be best placed to make this call.

**Review by anonymouse64 — APPROVED** [review]
lgtm, thanks for changing it to be a super-privileged as requested. We can merge this on Monday

**Comment by alexclewontin** [comment]
Fixed 2 tests in interfaces/policy/basedeclaration_test.go

File: interfaces/builtin/xilinx_dma.go
Comment (anonymouse64) [inline interfaces/builtin/xilinx_dma.go:51]: ah this is not a thing, you should just do

```suggestion
/sys/module/xdma/parameters/* r,
```

**Pull request merged**
